### PR TITLE
Use --file for relation-set; avoid shell argument length limits

### LINKED ISSUE #801: `RelationDataContent.__setitem__` should dynamically dispatch to a file if it's too long
We've already seen this with Grafana Dashboards, which routinely overflow the maximum argument length from `subprocess`, but it was also observed that relating Prometheus to a very large number of targets could overflow and cause a strange looking `OSError` on a `RelationChangedEvent`

Ultimately, this is due to [`relation_set`](https://github.com/canonical/operator/blob/0a097748299506c7651bdef99524638146f9724a/ops/model.py#L951) calling back to `subprocess` to handle [`relation-set ...`](https://github.com/canonical/operator/blob/0a097748299506c7651bdef99524638146f9724a/ops/model.py#L2136). 

We already [split long log messages](https://github.com/canonical/operator/pull/632), and `relation-set` takes a `--file` parameter which reads in YAML, allowing the limit to be bypassed. If OF determines that the length of the relation data is anywhere near the limit, we could defer to something like:

```python
with tempfile.TempFile() as relation_data:
    with open(relation_data, "w") as f:
        f.write(yaml.dump({key: value})
    self._backend.relation_set(..., data_file=relation_data)
```

If an optarg were added to `relation_set` where, if present, data was loaded from a file. This seems easy enough to add, avoids requiring charm authors to carefully think about the size/length of their data bags and potentially destructure them to avoid it mapping back to a `map[string]string` on the backend, and yields the desired behavior.

### PR DESCRIPTION
**Pull request by rbarry82**
## Please provide the following details to expedite review (and delete this heading)

In some cases, the value of any given relation data key may be too long to set within the boundaries of the maximum argument length of `subprocess` and/or shell invocation. Charm authors interact with this by setting `MutableMapping` data, and it is not obvious to authors that this maps back to a bare k:v set on the commandline, until the data exceeds some limit and an `OSError` appears in the logs on a `RelationEvent`.

Deferring to file-based input is already done for `state-set`. Match that pattern.

Closes #801 
## Checklist

 - [ ] Have any types changed? If so, have the type annotations been updated?
 - [ ] If this code exposes model data, does it do so thoughtfully, in a way that aids understanding?

       We want to avoid simple passthrough of model data, without contextualization, where possible.

 - [ ] Do error messages, if any, present charm authors or operators with enough information to solve the problem?

## QA steps

*Please replace with how we can verify that the change works.*

This modifies all `relation-set` operations, so integration tests cover it, though it can certainly be tested manually also with any operation on relation data.

## Documentation changes

No change to workflow or API, and this part of the backend doesn't need doc changes.

## Changelog

*Please include a bulleted list of items here, suitable for inclusion in a release changelog.*

- Use --file for relation-set, so we avoid shell argument length limits

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -114,6 +114,7 @@
 logger = logging.getLogger(__name__)
 
 MAX_LOG_LINE_LEN = 131071  # Max length of strings to pass to subshell.
+MAX_RELATION_SET_LINE_LEN = 131000  # Max length of relation key-value data
```
Comment (rwcarlsen) [inline ops/model.py:117]: It appears this is unused?

Comment (rbarry82) [reply ops/model.py]: It is -- forgotten from a first pass. Removed now.

File: ops/model.py [hunk 2]
```diff
@@ -2036,6 +2039,10 @@ def _run(self, *args: str, return_output: bool = False, use_json: bool = False
             args += ('--format=json',)
         try:
             result = run(args, **kwargs)
+
+            # pyright infers the first match when argument overloading/unpacking is used,
+            # so this needs to be coerced into the right type
+            result = typing.cast(CompletedProcess[bytes], result)
```
Comment (pengale) [inline ops/model.py:2045]: It looks like Python 3.8 is unhappy with this (see test output).

**Review by pengale — CHANGES_REQUESTED** [review]
I like the idea, but the failing unit tests need to be fixed before merging.

**Review by jnsgruk — APPROVED** [review]

**Pull request merged**